bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java

```diff
@@ -591,54 +659,89 @@ public void close() {
             closeSession();
             if (notifyHandler) {
                 bridgeHandler.onConnectionOffline();
             }
         }
     }
 
+    /**
+     * Close the event stream(s) if necessary.
+     */
+    private void closeEventStream() {
+        Session session = http2Session;
+        if (Objects.nonNull(session)) {
+            session.getStreams().stream().filter(s -> Objects.nonNull(s.getAttribute(EVENT_STREAM_ID)) && !s.isReset())
+                    .forEach(s -> {
+                        LOGGER.debug("closeEventStream() sessionId:{}, streamId:{}", session.hashCode(), s.getId());
+                        s.reset(new ResetFrame(s.getId(), ErrorCode.CANCEL_STREAM_ERROR.code), Callback.NOOP);
+                    });
+        }
+    }
+
     /**
      * Close the HTTP 2 session if necessary.
      */
     private void closeSession() {
-        LOGGER.debug("closeSession()");
         Session session = http2Session;
-        if (Objects.nonNull(session) && !session.isClosed()) {
-            session.close(0, null, Callback.NOOP);
+        if (Objects.nonNull(session)) {
+            LOGGER.debug("closeSession() sessionId:{}, openStreamCount:{}", session.hashCode(),
+                    session.getStreams().size());
+            session.close(ErrorCode.NO_ERROR.code, "closeSession", Callback.NOOP);
         }
         http2Session = null;
     }
 
     /**
-     * Method that is called back in case of fatal stream or session events. Note: under normal operation, the Hue
-     * Bridge sends a 'soft' GO_AWAY command every nine or ten hours, so we handle such soft errors by attempting to
-     * silently close and re-open the connection without notifying the handler of an actual 'hard' error.
+     * Close the given stream.
+     *
+     * @param stream to be closed.
+     */
+    private void closeStream(@Nullable Stream stream) {
+        if (Objects.nonNull(stream) && !stream.isReset()) {
+            stream.reset(new ResetFrame(stream.getId(), ErrorCode.NO_ERROR.code), Callback.NOOP);
+        }
+    }
+
+    /**
+     * Method that is called back in case of fatal stream or session events. The error is only processed if the
+     * connection is online, not in process of closing, and the identities of the current session and the session that
+     * caused the error are the same. In other words it ignores errors relating to expired sessions.
      *
      * @param listener the entity that caused this method to be called.
-     * @param cause the exception that caused the error.
+     * @param cause the type of exception that caused the error.
+     * @param sessionId the identity of the session on which the error occurred.
+     * @param streamId the identity of the stream on which the error occurred.
      */
-    private synchronized void fatalError(Object listener, Http2Exception cause) {
+    private synchronized void fatalError(Object listener, Http2Exception cause, int sessionId, int streamId) {
```

Comment:
Maybe pass a `Session` instead of its hash here and do the `hashCode()` within this function? Saves doing it in all callers.

Reply:
No. We have to release the hold on the Session object so it can be properly released. But we do need to know its id after it was so released.

Reply:
I was merely suggesting to pass it into the function, not to keep a permanent reference there ... basically just move the place that calls `.hashCode()`. You can't really control timing of destroying the object anyway.

Reply:
Ok. Got it. Fixed.